src/modules/concurrency-limit/ConcurrencyLimitModule.ts

```diff
@@ -27,36 +36,88 @@ export class ConcurrencyLimitModule extends BaseModule<IConcurrencyLimitConfig>
   }
 
   private async processSessionStart(session: FaucetSession): Promise<void> {
-    if(session.getSessionData<string[]>("skip.modules", []).indexOf(this.moduleName) !== -1)
+    if (
+      session
+        .getSessionData<string[]>("skip.modules", [])
+        .indexOf(this.moduleName) !== -1
+    )
       return;
     this.checkLimit(session);
   }
 
-  private checkLimit(session: FaucetSession): void {
-    if(this.moduleConfig.concurrencyLimit === 0)
+  private checkLimit(sessionStarting: FaucetSession): void {
+    const concurrentLimitByIP = this.moduleConfig.concurrencyLimitByIP;
+    const concurrentLimitByUserAndTargetAddress =
+      this.moduleConfig.concurrencyLimitByUserAndTargetAddress;
+
+    // Unlimited!
+    if (
+      concurrentLimitByIP === 0 ||
+      concurrentLimitByUserAndTargetAddress === 0
+    )
       return;
 
-    const activeSessions = ServiceManager.GetService(SessionManager).getActiveSessions();
+    // Values of session that we start
+    const userId = sessionStarting.getUserId();
+    const targetAddr = sessionStarting.getTargetAddr();
+    const remoteIP = sessionStarting.getRemoteIP();
+
+    const activeSessions = ServiceManager.GetService(SessionManager)
+      .getActiveSessions()
+      .filter((activeSession) => {
+        return activeSession !== sessionStarting;
+      });
 
-    const sessionsFromTheSameUser = activeSessions.filter((sess) => {
-      return sess !== session && (sess.getRemoteIP() === session.getRemoteIP() || sess.getTargetAddr() === session.getTargetAddr() || sess.getUserId() === session.getUserId())
-    });
+    // Limit by IP has been set
+    if (concurrentLimitByIP > 0) {
+      const sessions = activeSessions.filter((activeSession) => {
+        return activeSession.getRemoteIP() === remoteIP;
```

Comment:
There was a `sess !== session` check in original code and it might be a good idea to keep it just to be safe

Reply:
I moved it here: 
https://github.com/drpcorg/drpc-pow-faucet/pull/33/files#diff-cedc00ab489c67659bf7fd64c7378b8dd8c7c193004f6d3aea51ae031ddfc248R67-R69